### PR TITLE
Updated split encoders so indexes are based on left hand encoders first.

### PR DESCRIPTION
**Pull request by XScorpion2**
## Description

Updated split encoders so indexes are based on left hand encoders first.
This ensures when swapping master sides that code logic based on encoder index doesn't change.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/encoder.c [hunk 1]
```diff
@@ -60,12 +62,27 @@ void encoder_init(void) {
...
 }
 
 void encoder_read(void) {
   for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
     encoder_state[i] <<= 2;
     encoder_state[i] |= (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+#if SPLIT_KEYBOARD
+    encoder_value[i + thisHand] += encoder_LUT[encoder_state[i] & 0xF];
+    if (encoder_value[i + thisHand] >= ENCODER_RESOLUTION) {
+        encoder_update_kb(i, false);
```
Comment (XScorpion2) [inline quantum/encoder.c]: This should be i + thisHand

File: quantum/encoder.c [hunk 2]
```diff
@@ -60,12 +62,27 @@ void encoder_init(void) {
...
+    }
+    if (encoder_value[i + thisHand] <= -ENCODER_RESOLUTION) { // direction is arbitrary here, but this clockwise
+        encoder_update_kb(i, true);
```
Comment (XScorpion2) [inline quantum/encoder.c]: same here

File: quantum/encoder.c [hunk 3]
```diff
@@ -84,14 +102,14 @@ void encoder_state_raw(uint8_t* slave_state) {
 
 void encoder_update_raw(uint8_t* slave_state) {
   for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
-    encoder_value[NUMBER_OF_ENCODERS + i] += encoder_LUT[slave_state[i] & 0xF];
-    if (encoder_value[NUMBER_OF_ENCODERS + i] >= ENCODER_RESOLUTION) {
-        encoder_update_kb(NUMBER_OF_ENCODERS + i, false);
+    encoder_value[i + thatHand] += encoder_LUT[slave_state[i] & 0xF];
```
Comment (XScorpion2) [inline quantum/encoder.c]: This should be pulled out into a common function for this and the above code.

File: quantum/encoder.c (hunk 1 above)
Comment (XScorpion2) [reply quantum/encoder.c]: fixed

File: quantum/encoder.c (hunk 2 above)
Comment (XScorpion2) [reply quantum/encoder.c]: fixed

File: quantum/encoder.c (hunk 3 above)
Comment (XScorpion2) [reply quantum/encoder.c]: fixed

**Comment by XScorpion2** [comment]
All checks out swapping master halves on my Sol rev2. Encoder functionality remains in tact after swap.

**Comment by XScorpion2** [comment]
@zvecr Well, the only problem with the push method is that it the internal encoder arrays would still need to know to be 2x as large as well as which part of the array to modify based on left vs right logic when it is reading it's own state. Encoder support is actually closer to how matrix scanning works than rgblights, so with that being the case I would actually lean towards a separate split encoder file that is conditionally included over the basic encoder like what we do with the matrix. However, for the level of simple changes required for split encoder, I felt that was overkill.

**Comment by zvecr** [comment]
I wasn't suggesting that RGBLIGHT functionally was an identical use case, more an example of the model being used elsewhere. My main concern is definitely the fragmentation, for instance that RGB light code snippet is mixed with a fair amount internal logic within `quantum/rgblight.c`.

Maybe this is the best fix for right now, but I would at least like to consider the bigger picture.

**Comment by XScorpion2** [comment]
I got what you were saying, and I still think this is the more appropriate change for the reasons listed. I just don't think there is enough complexity or code in general for encoders to justify a different method.

If we push the data in from split common code into encoder, we would still need split logic in encoder to know that we need a 2x sized buffer, and that the encoder_read function would need to shift where it places its data. Which is most of the changes in this pr.

As I mentioned the other pattern is what we do for matrix scanning is that we have split and non-split matrix scanning code files. Personally I really dislike this as when someone wants to work on matrix code they have to update both files to keep them in sync. I've fallen into this trap a few times already where I've updated one and not the other.

**Comment by XScorpion2** [comment]
CI Failure was due to dns timeout.

**Comment by drashna** [comment]
@XScorpion2 could you test this again, since #6521 was merged?  Just want to make sure there is no unexpected behavior changes.

**Comment by XScorpion2** [comment]
Behavior remains unchanged

**Comment by XScorpion2** [comment]
Any word on additional reviews / merge?

**Comment by zvecr** [comment]
For transparency, internal discussions have largely been the same as #6480. Deciding if a behavioural change should go through the breaking change process.

**Comment by XScorpion2** [comment]
Fair enough, this pr does have a >0 impact as encoders are definitely frequently used and folks could have already gotten use to the existing odd behavior.

**Comment by XScorpion2** [comment]
Rebased on latest master, verified still working locally.
CI Failure is due to some unrelated issue in planck/rev1 default keymap.

**Pull request merged**